README.md
README file for the EXIF Drupal module.


Description
***********

The EXIF module allows to display EXIF metadata on image nodes. EXIF is a
specification for the image file format used by digital cameras.

The metadata tags defined in the EXIF standard cover a broad spectrum including:

 * Date and time information. Digital cameras will record the current date and
   time and save this in the metadata.
 * Camera settings. This includes static information such as the camera model
   and make, and information that varies with each image such as orientation,
   aperture, shutter speed, focal length, metering mode, and film speed
   information.
 * Location information, which could come from a GPS receiver connected to the
   camera.
 * Descriptions and copyright information.

Administrators can choose via CCK fields which EXIF information are read.

At this time, this module supports EXIF information only with JPEG files.

[1] Reference: http://en.wikipedia.org/wiki/Exchangeable_image_file_format


Requirements
**************

CCK with at least textfields enabled.
EXIF PHP Extension enabled.


Constraints
*************

Constraints with Image Processing utility :

Note that, it has been reported that when using GD, it strips some fields
(like field_gps_gpslatitude or field_gps_gpslongitude) from uploaded files.
Imagemagick seems not to have this limitation and does not strip these fields.
So Imagemagick is recommended when access to GPS data.

Usage
************
After installing it you can go to your content type. It supports both the image
module as well as the imagefield module. Let's say you have an content type
"photo". Go to your field settings and add a new field. For the name of the
field you need to follow the following naming conventions:

### Example:
#1 field_exif_exposuretime -> this would read the ExposureTime of the image and
save it in this field.

#2 field_ifd0_datetime ->  this would read the date time (2009:01:23 08:52:43)
of the image. as a field_type you can take for example a normal textfield, but
also a date field would be possible.

General rule is: [field]_[section]_[name] 

Under admin/settings/exif you can see a list of all possible information. These
informations are taken from the image "sample.jpg". I may not contain all tags
available. If you are looking for some specific tags you can just replace this
image with your own image.


CHANGELOG.txt
Exif 7.x-1.x-dev, xxxx-xx-xx
----------------------------
#3273954 by DamienMcKenna: Add CHANGELOG.txt.
#1336004 by DamienMcKenna, dman: Coding standards fixes.


Exif 7.x-1.10, 2022-04-06
-------------------------
#2046369 by martinjbaker: D7 support for exif_location.
#3008207: Incorrect APP1 Exif Identifier Code in Exif->readExifTags().
#2732203: Remove SXMP Library.
Fix some string not correctly converted.
#3268425 by DamienMcKenna: Add missing field admin permission for tests.
#3273794 by DamienMcKenna: Fix file permissions.
By FeyP, DamienMcKenna, mcdruid: Improved handling of Exif files.


Exif 7.x-1.9, 2018-03-21
------------------------
#2939127: No need to list exif.module in exif.info files[] list.
#2832511 by toasdt: Wrong watchdog() argument order.
#2825595 by spgd01: iptc multiple keywords creates blank terms.
#2773281 by makbuk: Warning: array_key_exists() expects parameter 2 to be array,
  null given in _exif_node_insert_update.
Bad retrieve spelling.
#2709967: support multi-valued image-fields.
#1918112: Data too long for column 'xxx' - We need to validate length of data
  before saving.
#1896498: Stream wrappers and performance.
#1723832: Further metadatas human-readable.
#2730897 by jphautin: Missing exif.class.php.


Exif 7.x-1.8, 2016-03-08
------------------------
#2681575: Behaviour of "Refresh on node update"


Exif 7.x-1.7, 2016-03-05
------------------------
ensure all vocabularies are loaded.


Exif 7.x-1.6, 2016-02-28
------------------------
Integration of exiftool.
#1692188: No exif data read until "Refresh on node update" setting is enabled.
#1917406: Notice: Undefined index: gpsaltituderef + Notice: Undefined index:
  gpsimgdirectionref.
#1981486 by firfin: drush exif-update always runs, even on nodetypes not
  selected.


Exif 7.x-1.5, 2016-02-28
------------------------
Issue #2659804: Getting exif data not working in 7.x-1.3


Exif 7.x-1.4, 2016-02-15
------------------------
Issue #1981660: Use of windows line endings
Issue #2550365: Image Processing with GD or Image Magic
Issue #2061249: Trim values before saving
Issue #1517118: 'Division by zero' error
Issue #2362237: Work with multiple vocabularies


Exif 7.x-1.3, 2016-02-28
------------------------
#2656744 by jphautin: Strange characters in Exif Tags.
#1445744: Existing term choosen instead of creating new one in the new
  vocabulary.
#1658502: Illegal string offset 'tag' in _exif_node_insert_update().
#1357038: i18n Taxonomy translation breaks exif term reference.
#2656744 by jphautin: Strange characters in Exif Tags.
#1289992 by Shawn DeArmond, firfin: Add computed_copyright, exif_lens,
  exif_lensid as selectable metadata.
#1458782 by ryne.andal: Alphabetical list of fields in select drop down.
#1019248 by marsdk, Xen: Incorrect APP1 Exif Identifier Code in
  Exif->readExifTags().
fix issue https://www.drupal.org/node/2325833
fix issue http://drupal.org/node/1366014
fix issue http://drupal.org/node/1529282
fix issue http://drupal.org/node/1956946
fix issue http://drupal.org/user/617182 media 2.0 API Change
update XMP integration to use XMP PHP Toolkit, 2.0-rev14
fix issue http://drupal.org/node/1529282


Exif 7.x-1.2, 2011-11-14
------------------------
Fix issue http://drupal.org/node/1297060
Fix issue http://drupal.org/node/1295832
Fix issue http://drupal.org/node/668688
Fix issue http://drupal.org/node/1289934
Fix issue http://drupal.org/node/1297028
Fix issue http://drupal.org/node/1277950
Fix issue http://drupal.org/node/1055750
Add first date module integration (see http://drupal.org/node/1191480).
Add first drush integration for batch processing (see
 http://drupal.org/node/668690). add first date module integration (see
 http://drupal.org/node/1191480).
Still working on date integration
First attempt to integrate with last version of date module.
Add media integration. can not apply patch as it make some regression).
Fix issue http://drupal.org/node/1206038
Fix issue http://drupal.org/node/1190328
Add functional test on gps and comment issues. add the image in the admin page.
Correct some mistake from the applied patch. still need some work as it seems
  there is regression with apostrophe character. :(
apply patch #12 from http://drupal.org/node/1176886
fix issue http://drupal.org/node/1189074


Exif 7.x-1.1, 2011-06-14
------------------------
Add a css to hidden term or textfield when cardinality of values is not set to
 1. correct apostrophe management with multiple text field.
 (related to http://drupal.org/node/1149740).
Added tests for non regression in simpletest suite.


Exif 7.x-1.0, 2011-06-13
------------------------
Added first simpletest suite to ensure no regression before first release issues
  fixed are : - http://drupal.org/node/1047220 (notice and warning message)
  - http://drupal.org/node/1149740 (apostrophe)
fix issue http://drupal.org/node/1019252. add new required library in the
  install requirements.
fix issue http://drupal.org/node/1019252 major rewriting of metadata values
  formatting.
fix issue http://www.drupal.org/node/1139944
major refactoring to manage media bundle. see  http://drupal.org/node/1055750
http://www.drupal.org/node/1113174
add .gitignore file
fix issue http://drupal.org/node/1084018
fix issue http://drupal.org/node/1052786
fix issue http://drupal.org/node/1065178
correction of issues: http://drupal.org/node/1052474
  http://drupal.org/node/1016092
Stripping CVS keywords
Update field setting management. adding image_field settings for each exif_field
Update install warning title and description correct creation/deletion of
  variable.
Update field management to be compliant with old mode (naming convention) or
  setup mode (widget field api).
#935220
First attempt to use widget field API
Format classes to follow drupal coding rules.
Update XMP integration
Merge with 6.x branch, version 1.3-alpha1 update to work on drupal 7.0-alpha5
Import of patch provided in issue 697244. see http://drupal.org/node/697244
#178370 added location exif

ExifFactory.php
<?php

namespace Drupal\exif;

include_once drupal_get_path('module', 'exif') . '/ExifPHPExtension.php';
include_once drupal_get_path('module', 'exif') . '/SimpleExiftoolFacade.php';

/**
 *
 */
class ExifFactory {

  /**
   *
   */
  public static function getExtractionSolutions() {
    return array(
      "simple_exiftool" => "exiftool",
      "php_extensions"  => "php extensions",
    );
  }

  /**
   *
   */
  public static function getExifInterface() {
    $extractionSolution = variable_get('exif_extraction_solution');
    $useExifToolSimple  = $extractionSolution == "simple_exiftool";
    if (isset($useExifToolSimple) && $useExifToolSimple && SimpleExifToolFacade::checkConfiguration()) {
      return SimpleExifToolFacade::getInstance();
    }
    else {
      // Default case for now (same behavior as previous versions).
      return ExifPHPExtension::getInstance();
    }
  }

}


ExifInterface.php
<?php

namespace Drupal\exif;

/**
 *
 */
interface ExifInterface {

  /**
   * Work out which fields have to be processed.
   *
   * Going through all the fields that have been created for a given node type
   * and try to figure out which match the naming convention -> so that we know
   * which exif information we have to read
   *
   * Naming convention are: field_exif_xxx (xxx would be the name of the exif
   * tag to read.
   *
   * @param array $fields
   *   The fields.
   *
   * @return array
   *   A list of EXIF tags to read for this image.
   */
  public function getMetadataFields(array $fields = array());

  /**
   * List of options for the method.
   *
   * @param string $file
   * @param bool $enable_sections
   *   Retrieve sections.
   *
   * @return array $data
   */
  public function readMetadataTags($file, $enable_sections = TRUE);

  /**
   *
   *
   * @return array
   */
  public function getFieldKeys();

}


SimpleExiftoolFacade.php
<?php

namespace Drupal\exif;

include_once drupal_get_path('module', 'exif') . '/ExifInterface.php';

/**
 * Helper class to handle the whole data processing of EXIF with exiftool.
 */
class SimpleExifToolFacade implements ExifInterface {

  static private $instance = NULL;

  /**
   *
   *
   * @return object
   */
  public static function getInstance() {
    if (is_null(self::$instance)) {
      self::$instance = new self;
    }
    return self::$instance;
  }

  public function getFieldKeys() {
    return array();
  }

  /**
   * Going through all the fields that have been created for a given node type
   * and try to figure out which match the naming convention -> so that we know
   * which exif information we have to read
   *
   * Naming convention are: field_exif_xxx (xxx would be the name of the exif
   * tag to read
   *
   * @param array $fields
   *   The fields to process.
   *
   * @return array
   *   A list of exif tags to read for this image.
   */
  public function getMetadataFields(array $fields = array()) {
    foreach ($fields as $drupal_field => $metadata_settings) {
      $metadata_field = $metadata_settings['metadata_field'];
      $ar = explode('_', $metadata_field);
      if (isset($ar[0])) {
        $section = $ar[0];
        unset($ar[0]);
        $fields[$drupal_field]['metadata_field'] = array(
          'section' => $section,
          'tag' => implode('_', $ar),
        );
      }
    }
    return $fields;
  }

  public static function checkConfiguration() {
    $exiftoolLocation = self::getExecutable();
    return isset($exiftoolLocation) && is_executable($exiftoolLocation);
  }

  public static function getExecutable() {
    return variable_get('exif_exiftool_location');
  }

  public function runTool($file, $enable_sections = TRUE, $enable_markerNote = FALSE, $enable_non_supported_tags = FALSE) {
    $params = ' -E -n -json ';
    if ($enable_sections) {
      $params .= '-g -struct ';
    }
    if ($enable_markerNote) {
      $params .= '-fast ';
    }
    else {
      $params .= '-fast2 ';
    }
    if ($enable_non_supported_tags) {
      $params .= '-u -U ';
    }

    // Escape all of the arguments passed to the function.
    // Note: If params is expanded so it is customizable, make sure that each
    // piece is passed through escapeshellarg().
    $commandline = escapeshellcmd('exiftool' . $params . escapeshellarg($file));
    $output = array();
    $returnCode = 0;
    exec($commandline, $output, $returnCode);

    if ($returnCode != 0) {
      $output = "";
      watchdog('exif', 'Exiftool returns an error. Can not extract metadata from file !file', array(
        '!file' => $file,
      ), WATCHDOG_WARNING);
    }
    return implode("\n", $output);
  }

  public function tolowerJsonResult(array $data) {
    $result = array();
    foreach ($data as $section => $values) {
      if (is_array($values)) {
        $result[strtolower($section)] = array_change_key_case($values);
      }
      else {
        $result[strtolower($section)] = $values;
      }
    }
    return $result;
  }

  public function readAllInformation($file, $enable_sections = TRUE, $enable_markerNote = FALSE, $enable_non_supported_tags = FALSE) {
    $jsonAsString = $this->runTool($file, $enable_sections, $enable_markerNote, $enable_non_supported_tags);
    $json = json_decode($jsonAsString, TRUE);
    $errorCode = json_last_error();
    if ($errorCode == JSON_ERROR_NONE) {
      return $this->tolowerJsonResult($json[0]);
    }
    else {
      $errorMessage = '';
      switch ($errorCode) {
        case JSON_ERROR_DEPTH:
          $errorMessage = 'Maximum stack depth exceeded';
          break;

        case JSON_ERROR_STATE_MISMATCH:
          $errorMessage = 'Underflow or the modes mismatch';
          break;

        case JSON_ERROR_CTRL_CHAR:
          $errorMessage = 'Unexpected control character found';
          break;

        case JSON_ERROR_SYNTAX:
          $errorMessage = 'Syntax error, malformed JSON';
          break;

        case JSON_ERROR_UTF8:
          $errorMessage = 'Malformed UTF-8 characters, possibly incorrectly encoded';
          break;

        default:
          $errorMessage = 'Unknown error';
          break;
      }
      // Logs a notice.
      watchdog('exif', 'Error reading information from exiftool for file !file: !message', array(
        '!file' => $file,
        '!message' => $errorMessage,
      ), WATCHDOG_NOTICE);
      return array();
    }
  }

  /**
   * $arOptions liste of options for the method :
   * # enable_sections : (default : TRUE) retrieve also sections.
   *
   * @param string $file
   * @param bool $enable_sections
   *
   * @return array $data
   */
  public function readMetadataTags($file, $enable_sections = TRUE) {
    if (!file_exists($file)) {
      return array();
    }
    $data = $this->readAllInformation($file, $enable_sections);
    return $data;
  }

  /**
   *
   *
   * @param array $metadata
   * @param array $tag_names
   *
   * @return array
   */
  public function filterMetadataTags(array $metadata, array $tag_names) {
    $info = array();
    foreach ($tag_names as $drupal_field => $metadata_settings) {
      $tagName = $metadata_settings['metadata_field'];
      if (!empty($metadata[$tagName['section']][$tagName['tag']])) {
        $info[$tagName['section']][$tagName['tag']] = $metadata[$tagName['section']][$tagName['tag']];
      }
    }
    return $info;
  }

}


exif_functionaltests.php
<?php

/**
 * Base class for functional tests for this module.
 */
abstract class ExifFunctionalTestCase extends DrupalWebTestCase {

  /**
   * A user account that has the appropriate permissions for the use case.
   *
   * @var object
   */
  protected $privilegedUser;

  /**
   * Create a new file field.
   *
   * @param string $name
   *   The name of the new field (all lowercase), exclude the "field_" prefix.
   * @param string $type_name
   *   The node type that this field will be added to.
   * @param array $field_settings
   *   A list of field settings that will be added to the defaults.
   * @param array $instance_settings
   *   A list of instance settings that will be added to the instance defaults.
   * @param array $widget_settings
   *   A list of widget settings that will be added to the widget defaults.
   */
  public function createFileField($name, $type_name, array $field_settings = array(), array $instance_settings = array(), array $widget_settings = array()) {
    $field = array(
      'field_name' => $name,
      'type' => 'file',
      'settings' => array(),
      'cardinality' => !empty($field_settings['cardinality']) ? $field_settings['cardinality'] : 1,
    );
    $field['settings'] = array_merge($field['settings'], $field_settings);
    field_create_field($field);

    $this->attachFileField($name, 'node', $type_name, $instance_settings, $widget_settings);
  }

  /**
   * Attach a file field to an entity.
   *
   * @param string $name
   *   The name of the new field (all lowercase), exclude the "field_" prefix.
   * @param string $entity_type
   *   The entity type this field will be added to.
   * @param string $bundle
   *   The bundle this field will be added to.
   * @param array $instance_settings
   *   A list of instance settings that will be added to the instance defaults.
   * @param array $widget_settings
   *   A list of widget settings that will be added to the widget defaults.
   */
  public function attachFileField($name, $entity_type, $bundle, array $instance_settings = array(), array $widget_settings = array()) {
    $instance = array(
      'field_name' => $name,
      'label' => $name,
      'entity_type' => $entity_type,
      'bundle' => $bundle,
      'required' => !empty($instance_settings['required']),
      'settings' => array(),
      'widget' => array(
        'type' => 'file_generic',
        'settings' => array(),
      ),
    );
    $instance['settings'] = array_merge($instance['settings'], $instance_settings);
    $instance['widget']['settings'] = array_merge($instance['widget']['settings'], $widget_settings);
    field_create_instance($instance);
  }

  /**
   * Get the fid of the last inserted file.
   */
  public function getLastFileId() {
    return (int) db_query('SELECT MAX(fid) FROM {file_managed}')->fetchField();
  }

  /**
   *
   */
  public function addNewFieldToContentType($contenttype, $fieldLabel, $fieldName, $fieldType, $fieldWidget, array $fieldSettings = array(), $settings = array()) {
    $edit = array(
      'fields[_add_new_field][label]' => $fieldLabel,
      'fields[_add_new_field][field_name]' => $fieldName,
      'fields[_add_new_field][type]' => $fieldType,
      'fields[_add_new_field][widget_type]' => $fieldWidget,
    );
    $this->drupalPost("admin/structure/types/manage/{$contenttype->type}/fields", $edit, 'Save');
    $this->drupalPost(NULL, $fieldSettings, 'Save field settings');
    $this->drupalPost(NULL, $settings, 'Save settings');
  }

  /**
   *
   */
  public function addExistingFieldToContentType($contenttype, $fieldLabel, $fieldName, $fieldWidget, array $fieldSettings = array()) {
    $edit = array(
      'fields[_add_existing_field][label]' => $fieldLabel,
      'fields[_add_existing_field][field_name]' => 'field_' . $fieldName,
      'fields[_add_existing_field][widget_type]' => $fieldWidget,
    );
    $this->drupalPost("admin/structure/types/manage/{$contenttype->type}/fields", $edit, 'Save');
    $this->drupalPost(NULL, $fieldSettings, 'Save settings');
  }

  /**
   *
   */
  public function manageDisplay($content_type, $fieldName, $label = 'hidden', $format = 'hidden', array $displays = array('full', 'teaser')) {
    $edit = array();
    // Accepted values are: above, inline, hidden.
    $edit["fields[field_{$fieldName}][label]"] = $label;
    // Accepted values for term are: taxonomy_term_reference_link,
    // taxonomy_term_reference_plain, hidden.
    // Accepted value for text are: text_plain, hidden.
    $edit["fields[field_{$fieldName}][type]"] = $format;
    foreach ($displays as $key => $display) {
      $edit["view_modes_custom[{$display}]"] = $display;
    }
    $this->drupalPost("admin/structure/types/manage/{$content_type->type}/display", $edit, 'Save');
  }

  /**
   *
   */
  public function activateMetadataOnObjectTypes(array $objectTypes = array(), $vid = 1) {
    if (!empty($objectTypes)) {
      $edit = array();
      foreach ($objectTypes as $key => $objectType) {
        $edit["exif_nodetypes[{$objectType}]"] = $objectType;
      }
      // Already existing vocabulaty tags - vid = 1.
      $edit['exif_vocabulary'] = $vid;
      $this->drupalPost('admin/config/media/exif/settings', $edit, 'Save configuration');
    }
  }

  /**
   *
   */
  public function setUp() {
    // Enable any modules required for the test.
    parent::setUp($this->initModules());
    // Types:
    // * image/Image
    // * taxonomy_term_reference/Term reference
    // * text/Text
    // Widgets
    // * exif_readonly/metadata from image
    // * image_image/Image
    // Create a "photo" content type.
    $settings = array(
      'type' => 'photo',
      'name' => 'Photography',
      'base' => 'node_content',
      'description' => 'show a photo and the metadata',
      'help' => '',
      'title_label' => 'Title',
      'body_label' => 'Body',
      'has_title' => 1,
      'has_body' => 0,
    );
    $type = $this->drupalCreateContentType($settings);

    $this->privileged_user = $this->drupalCreateUser(array(
      'administer image metadata',
      'administer content types',
      'administer nodes',
      'administer fields',
      'administer taxonomy',
      'create photo content',
      'edit any photo content',
    ));
    $this->assertNotEqual(FALSE, $this->privileged_user, 'user created.', 'Exif');
    $this->drupalLogin($this->privileged_user);

    $this->addExistingFieldToContentType($type, 'photo', 'image', 'image_image');
    $this->addNewFieldToContentType($type, 'model', 'exif_model', 'taxonomy_term_reference', 'exif_readonly');
    $this->addNewFieldToContentType($type, 'keywords', 'iptc_keywords', 'taxonomy_term_reference', 'exif_readonly', array(), array('field[cardinality]' => -1));
    $this->addNewFieldToContentType($type, 'usercomment', 'exif_usercomment', 'text', 'exif_readonly');
    $this->addNewFieldToContentType($type, 'latitude', 'gps_gpslatitude', 'text', 'exif_readonly');
    $this->addNewFieldToContentType($type, 'longitude', 'gps_gpslongitude', 'text', 'exif_readonly');
    $this->addNewFieldToContentType($type, 'latituderef', 'gps_gpslatituderef', 'text', 'exif_readonly');
    $this->addNewFieldToContentType($type, 'longituderef', 'gps_gpslongituderef', 'text', 'exif_readonly');
    $this->addNewFieldToContentType($type, 'artist', 'ifd0_artist', 'taxonomy_term_reference', 'exif_readonly');
    $this->addNewFieldToContentType($type, 'title', 'ifd0_title', 'text', 'exif_readonly');
    $this->addNewFieldToContentType($type, 'flash', 'exif_flash', 'text', 'exif_readonly');
    $this->addNewFieldToContentType($type, 'date', 'exif_datetime', 'text', 'exif_readonly');
    $this->addNewFieldToContentType($type, 'DateObject', 'ifd0_datetime', 'datetime', 'exif_readonly');
    $this->addNewFieldToContentType($type, 'fileDate', 'exif_filedatetime', 'text', 'exif_readonly');
    $this->addNewFieldToContentType($type, 'fileDateObject', 'ifd0_filedatetime', 'datetime', 'exif_readonly');

    $this->activateMetadataOnObjectTypes(array($type->type));

    $this->drupalGet("admin/structure/types/manage/photo/fields");
    $this->drupalGet("node/add");
    $this->assertResponse(200, t('User is allowed to add content types.'), 'Exif');
  }

  public function testCreatePhotoNodeWithoutImage() {
    $settings = array(
      'type' => 'photo',
      'title' => $this->randomName(32),
    );
    $node = $this->drupalCreateNode($settings);
    $this->assertNotNull($node, "node created", 'Exif');
    $this->verbose('Node created: ' . var_export($node, TRUE));
    $this->drupalGet("node/{$node->nid}/edit");
    $this->assertResponse(200, t('User is allowed to edit the content.'), 'Exif');
    $this->assertText(t("@title", array('@title' => $settings['title'])), "Found title in edit form", 'Exif');
    $this->drupalPost(NULL, array(), t('Save'));
    $this->assertResponse(200, t('trying to submit node.'), 'Exif');
    $this->assertNoText('The content on this page has either been modified by another user, or you have already submitted modifications using this form. As a result, your changes cannot be saved.', 'form has been correctly submitted', 'Exif');
    $this->drupalGet("node/{$node->nid}");
    $this->assertResponse(200, t('photography node edition is complete.'), 'Exif');
  }

  public function testCreatePhotoNodeWithImage() {
    $settings = array(
      'type' => 'photo',
      'title' => $this->randomName(32),
    );
    $node = $this->drupalCreateNode($settings);
    $this->assertNotNull($node, "node created", 'Exif');
    $this->verbose('Node created: ' . var_export($node, TRUE));
    $this->drupalGet("node/{$node->nid}/edit");
    // Check field settings are hidden.
    $this->assertRaw("exif.css");
    // Assert hidden field are present.
    $this->assertField("field_iptc_keywords[und][0][tid]", "hidden field keywords is present", "Exif");
    $this->assertField("field_exif_model[und][0][tid]", "hidden field model is present", "Exif");
    $this->assertResponse(200, t('User is allowed to edit the content.'), 'Exif');
    $this->assertText(t("@title", array('@title' => $settings['title'])), "Found title in edit form", 'Exif');
    $img_path = drupal_get_path('module', 'exif') . '/sample.jpg';
    $this->assertTrue(file_exists($img_path), "file {$img_path} exists.", "Exif");
    $file_upload = array(
      'files[field_image_und_0]' => $img_path,
    );
    $this->drupalPostAJAX(NULL, $file_upload, 'field_image_und_0_upload_button');
    $this->assertResponse(200, 'photo is uploaded.', 'Exif');
    $this->drupalGet("node/{$node->nid}/edit");
    $fid = $this->getLastFileId();
    $file_uploaded_attach = array(
      'field_image[und][0][fid]' => $fid,
    );
    $this->drupalPost(NULL, $file_uploaded_attach, 'Save');
    $this->assertResponse(200, t('trying to submit node.'), 'Exif');
    $this->assertNoText('The content on this page has either been modified by another user, or you have already submitted modifications using this form. As a result, your changes cannot be saved.', 'form has been correctly submitted', 'Exif');
    $this->drupalGet("node/{$node->nid}");
    $this->assertResponse(200, t('photography node edition is complete.'), 'Exif');

    // Check for label.
    $this->assertText("model");
    $this->assertText("keywords");
    $this->assertText("latitude");
    $this->assertText("longitude");
    $this->assertText("comment");
    $this->assertText("artist");
    $this->assertText("title");
    $this->assertText("flash");
    $this->assertText("date");
    $this->assertText("DateObject");
    $this->assertText("fileDate");
    $this->assertText("fileDateObject");

    // Check for values.
    $this->assertText("Canon EOS 350D DIGITAL", "model value is correct", "Exif");
    $this->assertText("annika", "keyword n°1 is correct", "Exif");
    $this->assertText("geburtstag", "keyword n°2 is correct", "Exif");
    $this->assertText("O&#039;Brien", "keyword n°3 is correct (apostrophe in text)", "Exif");
    $this->assertText("51.2977", "latitude is correct", "Exif");
    $this->assertText("12.2206", "longitude is correct", "Exif");
    $this->assertText("ich bin ein kleiner kommentar", "comment is correct", "Exif");
    $this->assertNoText("UNICODEich bin ein kleiner kommentar", "comment is correct", "Exif");
    $this->assertText("Raphael Schär", "artist is correct", "Exif");
    $this->assertText("Der Titel", "title is correct", "Exif");
    $this->assertText("Flash fired, compulsory flash mode", "flash is correct", "Exif");

    $this->assertText("2009-01-23T08:52:43", "date (wihtout Date module) is correct", "Exif");
    $this->assertText("2009/01/23 08:52:43", "date (with Date module) is correct", "Exif");

    $this->assertText("2011-11-05T21:39:10+01:00", "file date (wihtout Date module) is correct", "Exif");
    $this->assertText("2011/11/05 21:39:10", "file date (with Date module) is correct", "Exif");
  }

}
